openmicroscopy/openmicroscopy
# Lut webclient

# What this PR does

This adds Look-up Table (LUT) support in webclient, equivalent to #4764 in Insight.

# To test

This needs a full test of rendering settings, including selection of LUTs. Including:
 - Click colorpicker button, choose regular color (Red, Green etc) OR LUT OR pick with colorpicker.
 - Save, Undo/Redo, Apply To All
 - Copy / Paste of rendering settings (in Preview Pane and tree right-click menu)
 - Open Full image viewer (use button in Preview pane - current (unsaved) settings should be applied in full viewer) and test colorpicker & LUTs here too.

Webgateway API changes: cc @chris-allan 
I have added "lut" to the ```channelMarshal()``` so we have that in the ```imgData``` json if LUT is set:
```
channels: [
    {
        lut: "orange_hot.lut",
        color: "0000FF",
        emissionWave: 457,
        label: "DAPI",
        window: {
            max: 1323,
            end: 996,
            start: 172,
            min: 0
        },
        active: true
    },
```
For rendering images and elsewhere in the JavaScript code, I have used the LUT string in place of the color string since these are mutually exclusive:
E.g.  ``` ?c=1|172:996$16_colors.lut,2|23:1330$00FF00 ```
NB: LUT names containing ```, | $ :``` would break the parsing of this string. Will look into this more.

The Blitz Gateway changes handle this in ```image.setActiveChannels()``` where the ```colors``` parameter can include "cool.lut" strings as well as"FF0000".

ChannelWrapper has a ```getLut()``` which returns the name of the LUT.

Channel Buttons:
I have tried a couple of different approaches for rendering channel buttons with LUTs. If buttons are beside slider (which shows LUT) then maybe we don't need button to have LUT background since text is harder to read. However, in Image Viewer window, channel buttons aren't shown beside sliders. Here I have tried to use the centre of a "stretched" LUT as background, since this provides a smoother background color (allowing label to be read) which still giving an indication of the LUT. 
Thoughts?

![screen shot 2016-08-02 at 13 42 35](https://cloud.githubusercontent.com/assets/900055/17328772/72d675f8-58b7-11e6-8a38-98ca43ad8c2f.png)
